compiler/rustc_ast_lowering/src/lib.rs

```diff
@@ -1085,14 +1085,32 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
                 hir::TypeBindingKind::Equality { term }
             }
             AssocConstraintKind::Bound { bounds } => {
-                enum DesugarKind<'a> {
+                enum DesugarKind {
                     ImplTrait,
-                    Error(&'a ImplTraitPosition),
+                    Error(Option<Span>),
                     Bound,
                 }
 
                 // Piggy-back on the `impl Trait` context to figure out the correct behavior.
                 let desugar_kind = match itctx {
+                    ImplTraitContext::ReturnPositionOpaqueTy { .. }
+                    | ImplTraitContext::TypeAliasesOpaqueTy { .. }
+                    | ImplTraitContext::Universal
+                        if self.is_in_dyn_type =>
+                    {
+                        let bound_end_span = constraint
+                            .gen_args
+                            .as_ref()
+                            .map_or(constraint.ident.span, |args| args.span());
+                        let colon_span = if bound_end_span.eq_ctxt(constraint.span) {
+                            Some(self.tcx.sess.source_map().next_point(bound_end_span))
+                        } else {
+                            None
+                        };
+                        DesugarKind::Error(colon_span)
+                    }
+                    _ if self.is_in_dyn_type => DesugarKind::Error(None),
```

Comment:
nit: can you change this to first check `self.is_in_dyn_type` and then match on `itctx` inside of that? :sweat_smile: was a bit confusing for me as is